### PR TITLE
Python 3.11 Abi3 wheels

### PR DESCRIPTION
**Pull request by waahm7**
*Description of changes:*
- With a few minor changes, we can now have ABI3 compatible wheels for Python 3.11 and onward.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/event_stream_headers.c [hunk 1]
```diff
@@ -171,14 +171,15 @@ bool aws_py_event_stream_native_headers_init(struct aws_array_list *native_heade
...
+    const Py_ssize_t count = PySequence_Size(sequence_py);
     for (Py_ssize_t i = 0; i < count; ++i) {
         /* Borrowed reference, don't need to decref */
-        PyObject *header_py = PySequence_Fast_GET_ITEM(sequence_py, i);
+        PyObject *header_py = PySequence_GetItem(sequence_py, i);
 
         if (!s_add_native_header(native_headers, header_py)) {
             goto done;
         }
+        Py_XDECREF(header_py);
```
Comment (graebm) [inline source/event_stream_headers.c:182]: same 2 comments

File: source/http_stream.c [hunk 2]
```diff
@@ -121,7 +121,7 @@ static int s_on_incoming_header_block_done(
             goto done;
         }
 
-        PyList_SET_ITEM(header_list, i, tuple); /* steals reference to tuple */
+        PyList_SetItem(header_list, i, tuple); /* steals reference to tuple */
```
Comment (graebm) [inline source/http_stream.c:124]: this doesn't "steal a reference to the the tuple" anymore so this is a leak now

see, this is why it's good to comment on the lifetime stuff. If this comment wasn't here I wouldn't have remembered that

File: source/http_headers.c [hunk 3]
```diff
@@ -96,18 +96,19 @@ PyObject *aws_py_http_headers_add_pairs(PyObject *self, PyObject *args) {
         return NULL;
     }
 
-    const Py_ssize_t count = PySequence_Fast_GET_SIZE(py_sequence);
+    const Py_ssize_t count = PySequence_Size(py_sequence);
     for (Py_ssize_t i = 0; i < count; ++i) {
         /* XYZ_GET_ITEM() calls returns borrowed references */
-        PyObject *py_pair = PySequence_Fast_GET_ITEM(py_sequence, i);
+        PyObject *py_pair = PySequence_GetItem(py_sequence, i);
```
Comment (graebm) [inline source/http_headers.c:102]: delete comment about "returns borrowed references"
add comment about "New reference"

File: source/http_headers.c [hunk 4]
```diff
@@ -96,18 +96,19 @@ PyObject *aws_py_http_headers_add_pairs(PyObject *self, PyObject *args) {
...
 
-        if (!PyTuple_Check(py_pair) || PyTuple_GET_SIZE(py_pair) != 2) {
+        if (!PyTuple_Check(py_pair) || PyTuple_Size(py_pair) != 2) {
             PyErr_SetString(PyExc_TypeError, type_errmsg);
             goto done;
```
Comment (graebm) [inline source/http_headers.c:106]: if we take this branch, we leak

File: source/http_headers.c [hunk 5]
```diff
@@ -96,18 +96,19 @@ PyObject *aws_py_http_headers_add_pairs(PyObject *self, PyObject *args) {
...
         }
 
-        struct aws_byte_cursor name = aws_byte_cursor_from_pyunicode(PyTuple_GET_ITEM(py_pair, 0));
-        struct aws_byte_cursor value = aws_byte_cursor_from_pyunicode(PyTuple_GET_ITEM(py_pair, 1));
+        struct aws_byte_cursor name = aws_byte_cursor_from_pyunicode(PyTuple_GetItem(py_pair, 0));
+        struct aws_byte_cursor value = aws_byte_cursor_from_pyunicode(PyTuple_GetItem(py_pair, 1));
```
Comment (graebm) [inline source/http_headers.c:110]: These calls to PyTuple_GetItem() are creating new references that are leaking

File: source/mqtt5_client.c [hunk 6]
```diff
@@ -200,7 +200,7 @@ static PyObject *s_aws_set_user_properties_to_PyObject(
             Py_XDECREF(user_properties_list);
             return NULL;
         }
-        PyList_SET_ITEM(user_properties_list, i, tuple); /* Steals reference to tuple */
+        PyList_SetItem(user_properties_list, i, tuple); /* Steals reference to tuple */
```
Comment (graebm) [inline source/mqtt5_client.c:203]: yeah, all these changes are leaks, I'm going to stop commenting on every one

**Review by graebm — CHANGES_REQUESTED** [review]

File: setup.py [hunk 7]
```diff
@@ -283,6 +283,16 @@ def run(self):
...
 
 
+class bdist_wheel_abi3(bdist_wheel):
+    def get_tag(self):
+        python, abi, plat = super().get_tag()
+        if python.startswith("cp") and sys.version_info >= (3, 11):
+            # on CPython, our wheels are abi3 and compatible back to 3.11
+            return "cp311", "abi3", plat
+
+        return python, abi, plat
```
[the remark this answers is not in the document]
Comment (waahm7) [reply setup.py:293]: https://github.com/joerick/python-abi3-package-sample

File: source/http_headers.c [hunk 8]
```diff
@@ -175,8 +176,8 @@ static PyObject *s_py_tuple_from_header(struct aws_http_header header) {
         goto error;
     }
 
-    PyTuple_SET_ITEM(py_pair, 0, py_name);
-    PyTuple_SET_ITEM(py_pair, 1, py_value);
+    PyTuple_SetItem(py_pair, 0, py_name);
+    PyTuple_SetItem(py_pair, 1, py_value);
```
[the remark this answers is not in the document]
Comment (waahm7) [reply source/http_headers.c:189]: Discussed offline that PyTuple_SetItem steals a reference, so we should be fine in this case.

File: setup.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (graebm) [reply setup.py:293]: Cool, just curious

File: source/http_headers.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (graebm) [reply source/http_headers.c:189]: nevertheless, please add comments to all these calls noting whether it's a "New reference", "Steals reference", etc

File: source/auth_credentials.c [hunk 9]
```diff
@@ -526,11 +526,12 @@ PyObject *aws_py_credentials_provider_new_chain(PyObject *self, PyObject *args)
     }
 
     for (size_t i = 0; i < provider_count; ++i) {
-        PyObject *provider_py = PySequence_Fast_GET_ITEM(providers_pyseq, i);
+        PyObject *provider_py = PySequence_GetItem(providers_pyseq, i);
```
[the remark this answers is not in the document]
Comment (waahm7) [reply source/auth_credentials.c]: Thanks, fixed.

File: source/auth_credentials.c [hunk 10]
```diff
@@ -735,7 +736,7 @@ PyObject *aws_py_credentials_provider_new_cognito(PyObject *self, PyObject *args
             }
 
             for (size_t i = 0; i < logins_count; ++i) {
-                PyObject *login_tuple_py = PySequence_Fast_GET_ITEM(logins_pyseq, i);
+                PyObject *login_tuple_py = PySequence_GetItem(logins_pyseq, i);
```
[the remark this answers is not in the document]
Comment (waahm7) [reply source/auth_credentials.c]: Thanks, fixed.

File: source/auth_credentials.c [hunk 11]
```diff
@@ -752,6 +753,7 @@ PyObject *aws_py_credentials_provider_new_cognito(PyObject *self, PyObject *args
                         i);
                     goto done;
                 }
+                Py_XDECREF(login_tuple_py);
```
[the remark this answers is not in the document]
Comment (waahm7) [reply source/auth_credentials.c:757]: Thanks, fixed.

File: source/event_stream_headers.c (hunk 1 above)
Comment (waahm7) [reply source/event_stream_headers.c]: Thanks, fixed.

File: source/http_headers.c (hunk 3 above)
Comment (waahm7) [reply source/http_headers.c]: Thanks, fixed.

File: source/http_headers.c (hunk 4 above)
Comment (waahm7) [reply source/http_headers.c]: Discussed offline, that PyTuple steals a reference.

File: .github/workflows/ci.yml [hunk 12]
```diff
@@ -52,6 +52,7 @@ jobs:
           - cp39-cp39
           - cp310-cp310
           - cp311-cp311
+          - cp312-cp312
```
Comment (graebm) [inline .github/workflows/ci.yml:55]: do we need to update aws-crt-builder to be sure we're getting cp312 from the official docker images?

File: setup.py [hunk 13]
```diff
@@ -350,14 +362,20 @@ def awscrt_ext():
...
         '_awscrt',
         language='c',
         libraries=libraries,
         sources=glob.glob('source/*.c'),
         extra_compile_args=extra_compile_args,
         extra_link_args=extra_link_args,
-        extra_objects=extra_objects
+        extra_objects=extra_objects,
+        define_macros=macros,
+        py_limited_api=py_limited_api,
```
Comment (graebm) [inline setup.py:378]: utterly trivial: rename ~macros~ -> `define_macros`
so it lines up with this function's arg name, just like we're doing with all the other args

File: source/http_headers.c [hunk 14]
```diff
@@ -85,6 +85,29 @@ PyObject *aws_py_http_headers_add(PyObject *self, PyObject *args) {
...
+static int s_http_headers_add_pair(PyObject *py_pair, struct aws_http_headers *headers) {
+
+    const char *type_errmsg = "List of (name,value) pairs expected.";
+    if (!PyTuple_Check(py_pair) || PyTuple_Size(py_pair) != 2) {
+        PyErr_SetString(PyExc_TypeError, type_errmsg);
+        return AWS_OP_ERR;
+    }
+
+    struct aws_byte_cursor name = aws_byte_cursor_from_pyunicode(PyTuple_GetItem(py_pair, 0) /* Steals a reference */);
+    struct aws_byte_cursor value = aws_byte_cursor_from_pyunicode(PyTuple_GetItem(py_pair, 1) /* Steals a reference */);
```
Comment (graebm) [inline source/http_headers.c:97]: trivial:
```suggestion
    struct aws_byte_cursor name = aws_byte_cursor_from_pyunicode(PyTuple_GetItem(py_pair, 0) /* Borrowed reference */);
    struct aws_byte_cursor value = aws_byte_cursor_from_pyunicode(PyTuple_GetItem(py_pair, 1) /* Borrowed reference */);
```

File: source/http_headers.c [hunk 15]
```diff
@@ -85,6 +85,29 @@ PyObject *aws_py_http_headers_add(PyObject *self, PyObject *args) {
     Py_RETURN_NONE;
 }
 
+static int s_http_headers_add_pair(PyObject *py_pair, struct aws_http_headers *headers) {
+
+    const char *type_errmsg = "List of (name,value) pairs expected.";
+    if (!PyTuple_Check(py_pair) || PyTuple_Size(py_pair) != 2) {
+        PyErr_SetString(PyExc_TypeError, type_errmsg);
+        return AWS_OP_ERR;
```
Comment (graebm) [inline source/http_headers.c:93]: use python-style errors, or AWS-style errors, but don't mix them like this. It leads to people forgetting to set the "right" kind of thread-local state when errors happen, or assuming that aws_raise_error() was called when it wasn't, and stuff like that

It's not 100% consistent, but I've tried to put "py" in the name of functions that do python-style errors. I'd suggest reworking this like:

~static int s_http_headers_add_pair(...)~ -> `static bool s_py_http_headers_add_pair(...)`

File: source/mqtt5_client.c [hunk 16]
```diff
@@ -2030,23 +2036,35 @@ PyObject *aws_py_mqtt5_client_get_stats(PyObject *self, PyObject *args) {
         goto done;
     }
 
-    PyTuple_SET_ITEM(result, 0, PyLong_FromUnsignedLongLong((unsigned long long)stats.incomplete_operation_count));
-    if (PyTuple_GET_ITEM(result, 0) == NULL) {
+    PyTuple_SetItem(
+        result,
+        0,
+        PyLong_FromUnsignedLongLong((unsigned long long)stats.incomplete_operation_count)); /* Steals a reference */
+    if (PyTuple_GetItem(result, 0) == NULL) {                                               /* Steals a reference */
```
Comment (graebm) [inline source/mqtt5_client.c:2039]:
```suggestion
    if (PyTuple_GetItem(result, 0) == NULL) {                                               /* Borrowed reference */
```

File: source/mqtt5_client.c [hunk 17]
```diff
@@ -2030,23 +2036,35 @@ PyObject *aws_py_mqtt5_client_get_stats(PyObject *self, PyObject *args) {
...
         goto done;
     }
 
-    PyTuple_SET_ITEM(result, 1, PyLong_FromUnsignedLongLong((unsigned long long)stats.incomplete_operation_size));
-    if (PyTuple_GET_ITEM(result, 1) == NULL) {
+    PyTuple_SetItem(
+        result,
+        1,
+        PyLong_FromUnsignedLongLong((unsigned long long)stats.incomplete_operation_size)); /* Steals a reference */
+    if (PyTuple_GetItem(result, 1) == NULL) {                                              /* Steals a reference */
```
Comment (graebm) [inline source/mqtt5_client.c:2047]:
```suggestion
    if (PyTuple_GetItem(result, 1) == NULL) {                                              /* Borrowed reference */
```

File: source/mqtt5_client.c [hunk 18]
```diff
@@ -2030,23 +2036,35 @@ PyObject *aws_py_mqtt5_client_get_stats(PyObject *self, PyObject *args) {
...
         goto done;
     }
 
-    PyTuple_SET_ITEM(result, 2, PyLong_FromUnsignedLongLong((unsigned long long)stats.unacked_operation_count));
-    if (PyTuple_GET_ITEM(result, 2) == NULL) {
+    PyTuple_SetItem(
+        result,
+        2,
+        PyLong_FromUnsignedLongLong((unsigned long long)stats.unacked_operation_count)); /* Steals a reference */
+    if (PyTuple_GetItem(result, 2) == NULL) {                                            /* Steals a reference */
```
Comment (graebm) [inline source/mqtt5_client.c:2055]:
```suggestion
    if (PyTuple_GetItem(result, 2) == NULL) {                                            /* Borrowed reference */
```

File: source/mqtt5_client.c [hunk 19]
```diff
@@ -2030,23 +2036,35 @@ PyObject *aws_py_mqtt5_client_get_stats(PyObject *self, PyObject *args) {
...
         goto done;
     }
 
-    PyTuple_SET_ITEM(result, 3, PyLong_FromUnsignedLongLong((unsigned long long)stats.unacked_operation_size));
-    if (PyTuple_GET_ITEM(result, 3) == NULL) {
+    PyTuple_SetItem(
+        result,
+        3,
+        PyLong_FromUnsignedLongLong((unsigned long long)stats.unacked_operation_size)); /* Steals a reference */
+    if (PyTuple_GetItem(result, 3) == NULL) {                                           /* Steals a reference */
```
Comment (graebm) [inline source/mqtt5_client.c:2063]:
```suggestion
    if (PyTuple_GetItem(result, 3) == NULL) {                                           /* Borrowed reference */
```

File: source/mqtt_client_connection.c [hunk 20]
```diff
@@ -1329,23 +1329,35 @@ PyObject *aws_py_mqtt_client_connection_get_stats(PyObject *self, PyObject *args
         goto done;
     }
 
-    PyTuple_SET_ITEM(result, 0, PyLong_FromUnsignedLongLong((unsigned long long)stats.incomplete_operation_count));
-    if (PyTuple_GET_ITEM(result, 0) == NULL) {
+    PyTuple_SetItem(
+        result,
+        0,
+        PyLong_FromUnsignedLongLong((unsigned long long)stats.incomplete_operation_count)); /* Steals a reference */
+    if (PyTuple_GetItem(result, 0) == NULL) {                                               /* Steals a reference */
```
Comment (graebm) [inline source/mqtt_client_connection.c:1336]:
```suggestion
    if (PyTuple_GetItem(result, 0) == NULL) {                                               /* Borrowed reference */
```

File: source/mqtt_client_connection.c [hunk 21]
```diff
@@ -1329,23 +1329,35 @@ PyObject *aws_py_mqtt_client_connection_get_stats(PyObject *self, PyObject *args
...
         goto done;
     }
 
-    PyTuple_SET_ITEM(result, 1, PyLong_FromUnsignedLongLong((unsigned long long)stats.incomplete_operation_size));
-    if (PyTuple_GET_ITEM(result, 1) == NULL) {
+    PyTuple_SetItem(
+        result,
+        1,
+        PyLong_FromUnsignedLongLong((unsigned long long)stats.incomplete_operation_size)); /* Steals a reference */
+    if (PyTuple_GetItem(result, 1) == NULL) {                                              /* Steals a reference */
```
Comment (graebm) [inline source/mqtt_client_connection.c:1344]:
```suggestion
    if (PyTuple_GetItem(result, 1) == NULL) {                                              /* Borrowed reference */
```

File: source/mqtt_client_connection.c [hunk 22]
```diff
@@ -1329,23 +1329,35 @@ PyObject *aws_py_mqtt_client_connection_get_stats(PyObject *self, PyObject *args
...
         goto done;
     }
 
-    PyTuple_SET_ITEM(result, 2, PyLong_FromUnsignedLongLong((unsigned long long)stats.unacked_operation_count));
-    if (PyTuple_GET_ITEM(result, 2) == NULL) {
+    PyTuple_SetItem(
+        result,
+        2,
+        PyLong_FromUnsignedLongLong((unsigned long long)stats.unacked_operation_count)); /* Steals a reference */
+    if (PyTuple_GetItem(result, 2) == NULL) {                                            /* Steals a reference */
```
Comment (graebm) [inline source/mqtt_client_connection.c:1352]:
```suggestion
    if (PyTuple_GetItem(result, 2) == NULL) {                                            /* Borrowed reference */
```

File: source/mqtt_client_connection.c [hunk 23]
```diff
@@ -1329,23 +1329,35 @@ PyObject *aws_py_mqtt_client_connection_get_stats(PyObject *self, PyObject *args
...
         goto done;
     }
 
-    PyTuple_SET_ITEM(result, 3, PyLong_FromUnsignedLongLong((unsigned long long)stats.unacked_operation_size));
-    if (PyTuple_GET_ITEM(result, 3) == NULL) {
+    PyTuple_SetItem(
+        result,
+        3,
+        PyLong_FromUnsignedLongLong((unsigned long long)stats.unacked_operation_size)); /* Steals a reference */
+    if (PyTuple_GetItem(result, 3) == NULL) {                                           /* Steals a reference */
```
Comment (graebm) [inline source/mqtt_client_connection.c:1360]:
```suggestion
    if (PyTuple_GetItem(result, 3) == NULL) {                                           /* Borrowed reference */
```

File: .github/workflows/ci.yml (hunk 12 above)
Comment (waahm7) [reply .github/workflows/ci.yml:55]: Thanks, updated.

File: source/http_headers.c (hunk 15 above)
Comment (waahm7) [reply source/http_headers.c]: Thanks, fixed.

**Pull request merged**